Add `rpm-ostree ex rebuild` command

jlebon opened the pull request:
This command reads treefile dropins in `/etc` and applies it to the
system. Only the OSTree container flow is supported for now. For the
client-side flow, see coreos#2326.

In the container flow, only `packages` is currently supported. But this
paves the way for supporting more derivation fields as well as making
currently "base" fields only become valid derivation fields too (by
promoting it from the `BaseComposeConfigFields` struct to
`TreeComposeConfig`).

What we're doing here is providing a "container" backend for derivation
treefiles, but coreos#2326 will provide a "client" backend which uses the core
more fully. But the inputs themselves are the exact same, hence
providing symmetry between the two flows. (See also discussions about
this in coreos/fedora-coreos-tracker#1054).

This will also allow us to drop the `microdnf` dependency which will be
done in a following patch.

jlebon commented:
Sample output:

```
bash-5.1# rpm -q ltrace
package ltrace is not installed

bash-5.1# cat tmp.yaml
packages:
  - ltrace

bash-5.1# rpm-ostree compose container tmp.yaml
Enabled rpm-md repositories: fedora-cisco-openh264 fedora-modular updates-modular updates fedora updates-archive
Importing rpm-md... done
rpm-md repo 'fedora-cisco-openh264' (cached); generated: 2021-09-21T18:07:30Z solvables: 4
rpm-md repo 'fedora-modular' (cached); generated: 2021-10-26T05:08:36Z solvables: 1283
rpm-md repo 'updates-modular' (cached); generated: 2022-01-06T01:57:22Z solvables: 1371
rpm-md repo 'updates' (cached); generated: 2022-01-18T01:34:10Z solvables: 17188
rpm-md repo 'fedora' (cached); generated: 2021-10-26T05:31:27Z solvables: 65732
rpm-md repo 'updates-archive' (cached); generated: 2022-01-18T02:08:37Z solvables: 20988
Resolving dependencies... done
Will download: 1 package (140.0?kB)
Downloading from 'fedora'... done
Installing: ltrace-0.7.91-42.fc35.x86_64 (fedora)

bash-5.1# rpm -q ltrace
ltrace-0.7.91-42.fc35.x86_64
```

cgwalters commented on src/libpriv/rpmostree-core-private.h:
```diff
@@ -32,8 +32,10 @@ G_BEGIN_DECLS
+  /* Whether we were created with new_system() or new_container() */
   gboolean is_system;
+  /* Whether we were created with new_container() */
+  gboolean is_container;
```
OK as is but I think we should switch to an enum at some point.

cgwalters commented on src/libpriv/rpmostree-core.cxx:
```diff
@@ -154,10 +154,8 @@ set_rpm_macro_define (const char *key, const char *value)
 RpmOstreeContext *
 rpmostree_context_new_base (OstreeRepo   *repo)
 {
-  g_assert (repo != NULL);
```
This part is going to be...interesting.  Heavy intersection with https://github.com/ostreedev/ostree-rs-ext/issues/159

cgwalters commented on src/app/rpmostree-compose-builtin-tree.cxx:
```diff
@@ -1661,3 +1661,93 @@ rpmostree_compose_builtin_extensions (int             argc,
 }
+
+static void
+state_action_changed_cb (DnfState       *state,
```
Hmm...I think we could actually just do this by default in `rpmostree-core.cxx` right?  I believe we bypass this section of libdnf in the ostree-native case.  That way it'd work out of the box in other code paths too.

cgwalters commented:
I find calling this `compose container` a bit confusing given https://github.com/coreos/rpm-ostree/issues/3238 which was actually about doing this *outside* a container.  

Maybe...`rpm-ostree configure /tmp.yaml`?  Per https://github.com/coreos/rpm-ostree/issues/2326 again there's no strong reason we can't support this as well "per machine" right?  So supporting a generic `configure` verb would then lead naturally to having the CLI drop the yaml file in `/etc/rpm-ostree/config.d/<hash of yaml>.yml` and just invoking `rpm-ostree rebuild` or so.

cgwalters reviewed (commented):
Prep patches seem sane, let's split them to a separate PR and get them in.

jlebon commented:
Prep patches split out in https://github.com/coreos/rpm-ostree/pull/3341.

cgwalters commented:
> I find calling this compose container a bit confusing given #3238 which was actually about doing this outside a container.

I was thinking about this more, and I realized even though we've been working on the same project for years now, we may view things differently here.

Today, `rpm-ostree compose` has two qualities: It's declarative, and it's about building a new thing, distinct from what's running; i.e. distinct from the "host" (or container).

To me, `compose` was always much more about the latter i.e. "cross builds" or "new base images" we might say.  Having it be declarative was just an obviously necessary thing in order to have a sane build system.

But I realized perhaps you've seen it more this whole time as being the *declarative* frontend, and today it just happens to only support cross builds.  Is that the case?

cgwalters commented:
> The reason I ended up under rpm-ostree compose for this was because that command is purely associated with the server-side: the CoreOS layering work makes the distinction a bit fuzzy, but it's still more of a server-side operation, 

But see above - I was talking about `compose` being about *cross builds*, which also happens to be how one wants to do it on the server side.

It's worth noting here of course that quite commonly today `rpm-ostree compose tree` itself normally runs inside a container (or in the case of coreos-assembler in production as you know, a supermin VM which is more like a kata container).

To state this a different way, the fact that we're mutating the current root is what I'm keying in on.

Now, I don't want to block this!  But I think this discussion isn't just bikeshedding - it's important to have a shared consensus and understanding of what these different CLI verbs mean, because it has a large impact on the system architecture.

jlebon commented:
> To state this a different way, the fact that we're mutating the current root is what I'm keying in on.

Right yup. I do agree that fact is more pertinent than the server/client distinction. I wouldn't read too much into the `compose` part. It's more about getting it out of where client-side users go looking.

Anyway, I've moved this to `rpm-ostree ex rebuild` now and made the command hidden. This turns into an opportunity to test the UX, although just as part of the container flow.

OK, got carried away a bit and also added support for reading dropins and completely ripping out microdnf. Can split some of this into separate PRs if preferred.

jlebon commented:
Split out the first four into https://github.com/coreos/rpm-ostree/pull/3350.

cgwalters commented on src/libpriv/rpmostree-container.cxx:
```diff
@@ -60,3 +60,16 @@ rpmostree_container_rebuild (rpmostreecxx::Treefile &treefile,
+{
+  // XXX: awkward: add a e.g. `treefile_new_client_from_fields()`
+  g_autofree char *joined = g_strjoinv ("\",\"", packages);
+  g_autofree char *treefile_s = g_strdup_printf ("{\"packages\": [\"%s\"]}", joined);
```
Eeek.  I think today we can do `treefile_new_empty()` and then let's add a `set_packages()` API?

jlebon replied:
Let's keep that in a follow-up? I'd like to consider something more generic than per-field methods.

cgwalters reviewed (commented):
Overall looks sane, a few comments.

jlebon commented on Cargo.toml:
```diff
@@ -78,6 +78,7 @@ tracing-subscriber = { version = "0.3", features = ["env-filter"] }
 tokio = { version = "1.15.0", features = ["full"] }
 xmlrpc = "0.15.1"
 termcolor = "1.1.2"
+glob = "0.3.0"
```
Yeah, coreos-installer uses it too, which is where I learned it from.

jlebon commented on rust/src/treefile.rs:
```diff
@@ -46,6 +47,9 @@ const INCLUDE_MAXDEPTH: u32 = 50;
 const COMPOSE_JSON_PATH: &str = "usr/share/rpm-ostree/treefile.json";
 
+/// Globby path to client-side treefiles.
+const CLIENT_TREEFILES_GLOB: &str = "/etc/rpm-ostree/origin.d/*.yaml";
```
I initially wanted to use the `**` bit, but agreed it's a bit overkill as is.

jlebon commented:
OK, added a test for this now. Currently fighting issues with cosa locally so I didn't actually verify that it passes, but we'll see what CI says.

cgwalters commented on rust/src/treefile.rs:
```diff
@@ -2260,3 +2271,45 @@ pub(crate) fn treefile_new_client(filename: &str, basearch: &str) -> CxxResult<B
+pub(crate) fn treefile_new_client_from_etc(basearch: &str) -> CxxResult<Box<Treefile>> {
+    let basearch = opt_string(basearch);
+    let mut cfg = TreeComposeConfig::default();
+    let mut tfs = iter_etc_treefiles()?.collect::<Result<Vec<PathBuf>>>()?;
+    tfs.sort(); // sort because order matters; later treefiles override earlier ones
```
Oh right, that's a trap.  Hmm.  Also an argument for having the API for loading config files return a `Vec<>` directly.  But that can be a followup.

Or...I dunno actually, the more I think about this the more I think we shouldn't allow overrides at all?  Allowing it just seems like it's most likely to be a footgun for users.

At least this isn't a problem yet right because we're only supporting packages which are always unionioned?

cgwalters reviewed (approved):
Awesome work here!
